### ISSUE TITLE
Porting and hacking Llama2.c for fun and profit

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#11892]
Porting and hacking Llama2.c for fun and profit<br>
<br>
https://ift.tt/U61RxPK<br>
<br>
<br>
<br>

<div><div><div id="readability-page-1"><section><p>Over the last few weeks I’ve been playing around with the excellent <a href="https://github.com/karpathy/llama2.c">llama2.c</a> repository, which is basically a simple one-file C implementation of the Transformer architecture that powers models like Meta’s LLaMA and OpenAI’s GPT models. I had a great time trying to <a href="https://github.com/sgoedecke/llama2.js">port</a> llama2.c into a higher-level language (I tried Ruby and was ultimately successful with Javascript).</p>
<p>I ended up hacking a few interesting behaviours into the attention and token-generating mechanisms. This post will go through my experience, what I found difficult, what was surprisingly easy, and what I played around with once I had a working local implementation. I’ll assume familiarity with how LLMs work (see my <a href="https://www.seangoedecke.com/how-llms-work">post on that</a> if you want).</p>
<p>Overall I’d say the tokenizing was more straightforward than I thought it’d be, and decoding was <em>much</em> more straightforward. I wasn’t aware of the token scores and token merging algorithm, but it makes a lot of sense. Parsing the <code>.bin</code> model file was fine too. The hardest part was my choice to use actual multi-dimensional tensors for the weights, instead of copying the C implementation’s use of regular 1-D arrays with offsets all over the place. I reviewed a couple of higher-level language ports for inspiration, but they all used offsets too, and I’d hacked most of it together before I noticed that the C++ <a href="https://github.com/leloykun/llama2.cpp/blob/master/run.cpp">port</a> (of all things) used multi-dimensional tensors. Thanks to @leloykun, that really helped me pull it together.</p>
<p>My first few iterations of this generated repeated tokens (e.g. the prompt “Once” produced “Once cececececececece”). Debugging this was a <em>giant</em> pain, and convinced me that big AI labs must have a bunch of internal model debugging tools (or there’s something in PyTorch that I’m not aware of). Since I had a model C implementation at hand, I ended up dialing the temp down to 0 on both, using the same prompt, and stepping through the code in both programs to see where my program went wrong. Since this meant comparing very large arrays of 32-bit floats, my strategy was to print the first five floats of the array I wanted to compare and eyeball it. Eventually I graduated to a few larger randomly-chosen indexes, to give myself a better chance of catching problems in heads that weren’t head 0 (since at least some of the logic only kicks in then.)</p>
<p>I had to fix two or three silly bugs (e.g. typing <code>xb2</code> instead of <code>xb</code>) before the model started generating plausible sentences. It still wasn’t the same as the <code>llama2.c</code> output, though: after eight or so tokens it began to drift. Interestingly, the difference felt <em>meaningful</em>. The C implementation told a story that related to the prompt, while mine didn’t: whatever the prompt was, it began telling a generic story that had no relation. I felt this had to be a problem with the attention mechanism, since it was generating a coherent output but wasn’t paying attention to my prompt - and I was right! I was using an attention array that was far too small, only large enough to hold a single head’s worth of attention:</p>
<div><pre><code>    <span>att</span><span>:</span> <span>new</span> <span>Float32Array</span><span>(</span>config<span>.</span>seqLen<span>)</span></code></pre></div>
<p> I fixed that by changing it to:</p>
<div><pre><code>   <span>att</span><span>:</span> <span>new</span> <span>Float32Array</span><span>(</span>config<span>.</span>seqLen <span>*</span> config<span>.</span>nHeads<span>)</span></code></pre></div>
<p>Now it finally produced identical output to the C implementation.</p>
<p>Speaking of 32-bit floats, my first attempt was to write this in Ruby. I got much of the way there before realising that Ruby only supports 64-bit, which demoralized me enough to stop and choose JavaScript. I don’t actually know if a 64-bit implementation could work, or if the slight differences in calculations would screw up the weights. My guess is it would work - if you can quantize down to 4-bit ints, surely you can go the other way - but it does make it harder to debug since you’re not comparing exactly identical floats.</p>
<p>Once I had a working end-to-end implementation, I had two goals: to dig in and try and develop an intuitive understanding of transformers, and to play around and fiddle with some ideas I had. If you want to read my attempt at that first goal, see my <a href="https://www.seangoedecke.com/how-llms-work">post on that</a>. The first thing I wanted to fiddle around with was to try and visualize the current attention weights. There were two steps to this. First, since attention weights are separate per-head, I had to combine them into a single number after the token was generated:</p>
<div><pre><code><span>let</span> mashedAtt <span>=</span> runState<span>.</span>att<span>.</span><span>subarray</span><span>(</span><span>0</span><span>,</span> config<span>.</span>seqLen<span>)</span>
<span>for</span> <span>(</span><span>let</span> head <span>=</span> <span>1</span><span>;</span> head <span>&lt;</span> config<span>.</span>nHeads<span>;</span> head<span>++</span><span>)</span> <span>{</span>
    <span>const</span> att <span>=</span> runState<span>.</span>att<span>.</span><span>subarray</span><span>(</span>head <span>*</span> config<span>.</span>seqLen<span>,</span> <span>(</span>head <span>+</span> <span>1</span><span>)</span> <span>*</span> config<span>.</span>seqLen<span>)</span>
    <span>for</span> <span>(</span><span>let</span> i <span>=</span> <span>0</span><span>;</span> i <span>&lt;</span> att<span>.</span>length<span>;</span> i<span>++</span><span>)</span> <span>{</span>
        mashedAtt<span>[</span>i<span>]</span> <span>+=</span> att<span>[</span>i<span>]</span>
    <span>}</span>
<span>}</span></code></pre></div>
<p>Then I assigned <code>mashedAtt</code> to an object that held the token (which I called <code>piece</code>), and displayed it in the terminal like so:</p>
<div><pre><code><span>function</span> <span>printWithHighlightedAttention</span><span>(</span><span>output</span><span>)</span> <span>{</span>
    console<span>.</span><span>clear</span><span>(</span><span>)</span> 

    buf <span>=</span> <span>''</span>
    output<span>.</span><span>forEach</span><span>(</span><span>(</span><span>chunk<span>,</span> i</span><span>)</span> <span>=&gt;</span> <span>{</span>
        <span>let</span> colour
        <span>const</span> currentAttention <span>=</span> output<span>[</span>output<span>.</span>length <span>-</span> <span>1</span><span>]</span><span>.</span>att<span>[</span>i<span>]</span>
        <span>if</span> <span>(</span>currentAttention <span>&gt;</span> <span>0.7</span><span>)</span> <span>{</span>
            colour <span>=</span> colours<span>.</span>bright
        <span>}</span> <span>else</span> <span>if</span> <span>(</span>currentAttention <span>&gt;</span> <span>0.1</span><span>)</span> <span>{</span>
            colour <span>=</span> colours<span>.</span>reset
        <span>}</span> <span>else</span> <span>{</span>
            colour <span>=</span> colours<span>.</span>dim
        <span>}</span>
        buf <span>+=</span> colour
        buf <span>+=</span> chunk<span>.</span>piece
        buf <span>+=</span> colours<span>.</span>reset
    <span>}</span><span>)</span>
    process<span>.</span>stdout<span>.</span><span>write</span><span>(</span>buf<span>)</span>
<span>}</span></code></pre></div>
<p>Note that each token changes the attention scores for each previous token. Being able to visualize this helps develop an intuition of what the attention logic is doing.</p>
<p><span>
      <a href="https://www.seangoedecke.com/static/d852ee68c3714d8325194e0175a56d1b/c0566/attention.png" target="_blank" rel="noopener">
    <span></span>
  <img alt="attention" title="attention" src="https://www.seangoedecke.com/static/d852ee68c3714d8325194e0175a56d1b/fcda8/attention.png"></a>
    </span></p>
<p>The token we’ve just generated is <code>He</code>. Note that the most relevant tokens here are the previous pronoun <code>He</code> and the name <code>John</code> - that’s the model attempting to be consistent between names and pronouns. It’s not relevant that he’s a raven, or that he went to the park, or that he’s excited. Continuing the story, take this later screenshot:</p>
<p><span>
      <a href="https://www.seangoedecke.com/static/34879eac2bb5521d4f4519d9ac677c14/ddc6c/attention2.png" target="_blank" rel="noopener">
    <span></span>
  <img alt="attention2" title="attention2" src="https://www.seangoedecke.com/static/34879eac2bb5521d4f4519d9ac677c14/fcda8/attention2.png"></a>
    </span></p>
<p>Now the token we’ve generated is <code>birds</code>. Note that <code>raven</code> is now highlighted - the fact that John is a raven is now something the model needs to generate the next token.</p>
<p>I won’t share screenshots of this since it’s common in other implementations, but it’s equally trivial to visualize the confidence with which the model selects each token and highlight that:</p>
<div><pre><code>
<span>const</span> topLogits <span>=</span> Array<span>.</span><span>from</span><span>(</span>runState<span>.</span>logits<span>)</span><span>.</span><span>sort</span><span>(</span><span>(</span><span>a<span>,</span> b</span><span>)</span> <span>=&gt;</span> b <span>-</span> a<span>)</span><span>.</span><span>slice</span><span>(</span><span>0</span><span>,</span> <span>3</span><span>)</span>
<span>const</span> topLogitDiff <span>=</span> topLogits<span>[</span><span>0</span><span>]</span> <span>-</span> topLogits<span>[</span><span>2</span><span>]</span>
<span>...</span>
<span>function</span> <span>printWithHighlightedConfidence</span><span>(</span><span>output</span><span>)</span> <span>{</span>
    <span>const</span> lastChunk <span>=</span> output<span>[</span>output<span>.</span>length <span>-</span> <span>1</span><span>]</span>
    <span>let</span> colour
    <span>let</span> buf <span>=</span> <span>''</span>
    <span>const</span> diff <span>=</span> lastChunk<span>.</span>topLogitDiff
    <span>if</span> <span>(</span>diff <span>&gt;</span> <span>6</span><span>)</span> <span>{</span>
        colour <span>=</span> colours<span>.</span>bright
    <span>}</span> <span>else</span> <span>if</span> <span>(</span>diff <span>&gt;</span> <span>2</span><span>)</span> <span>{</span>
        colour <span>=</span> colours<span>.</span>reset
    <span>}</span> <span>else</span> <span>{</span>
        colour <span>=</span> colours<span>.</span>dim
    <span>}</span>
    buf <span>+=</span> colour
    buf <span>+=</span> lastChunk<span>.</span>piece
    buf <span>+=</span> colours<span>.</span>reset
    process<span>.</span>stdout<span>.</span><span>write</span><span>(</span>buf<span>)</span>
<span>}</span></code></pre></div>
<p>One other thing I did that I thought was interesting is <em>update</em> the attention mechanism in-flight. My idea was to accept both a prompt and an “emphasis” substring of the prompt, which my code would artifically inflate the attention values of. I know there’s something similar in AI visual output prompting, but I hadn’t seen anything like this in textual output, so I was excited to try it out.</p>
<p>The first step, which proved surprisingly tricky, was identifying the position of the emphasis substring in the prompt. Since prompts aren’t tokenized by words, and are tokenized in a way that depends on token <em>score</em>, I couldn’t simply tokenize the emphasis and expect it to be a sub-array of the prompt tokens. The emphasis word might be tokenized differently in the prompt itself. There’s definitely more efficient ways of doing this, but I ended up iterating over the prompt tokens twice with a sliding head/tail:</p>
<div><pre><code>
<span>const</span> emphasisPos <span>=</span> <span>{</span> <span>left</span><span>:</span> <span>0</span><span>,</span> <span>right</span><span>:</span> <span>0</span> <span>}</span>
<span>for</span> <span>(</span><span>let</span> i <span>=</span> <span>0</span><span>;</span> i <span>&lt;</span> promptTokens<span>.</span>length<span>;</span> i<span>++</span><span>)</span> <span>{</span>
    <span>const</span> head <span>=</span> promptTokens<span>.</span><span>slice</span><span>(</span>i<span>)</span><span>.</span><span>map</span><span>(</span><span>(</span><span>token</span><span>)</span> <span>=&gt;</span> tokenizer<span>.</span>vocab<span>[</span>token<span>]</span><span>)</span><span>.</span><span>join</span><span>(</span><span>""</span><span>)</span>
    <span>if</span> <span>(</span><span>!</span>head<span>.</span><span>includes</span><span>(</span>emphasis<span>)</span><span>)</span> <span>{</span>
        emphasisPos<span>.</span>left <span>=</span> i <span>-</span><span>1</span>
        <span>break</span>
    <span>}</span>
<span>}</span>
<span>for</span> <span>(</span><span>let</span> i <span>=</span> promptTokens<span>.</span>length<span>;</span> i<span>&gt;</span><span>0</span><span>;</span> i<span>--</span><span>)</span> <span>{</span>
    <span>const</span> tail <span>=</span> promptTokens<span>.</span><span>slice</span><span>(</span><span>-</span>i<span>)</span><span>.</span><span>map</span><span>(</span><span>(</span><span>token</span><span>)</span> <span>=&gt;</span> tokenizer<span>.</span>vocab<span>[</span>token<span>]</span><span>)</span><span>.</span><span>join</span><span>(</span><span>""</span><span>)</span>
    <span>if</span> <span>(</span><span>!</span>tail<span>.</span><span>includes</span><span>(</span>emphasis<span>)</span><span>)</span> <span>{</span>
        emphasisPos<span>.</span>right <span>=</span> promptTokens<span>.</span>length <span>-</span> i <span>+</span> <span>1</span>
        <span>break</span>
    <span>}</span>
<span>}</span></code></pre></div>
<p>Then, inside the part of the forward-pass step that generated the attention scores, I added a magic emphasis factor:</p>
<div><pre><code>
<span>for</span> <span>(</span><span>let</span> j <span>=</span> <span>0</span><span>;</span> j <span>&lt;=</span> att<span>.</span>length<span>;</span> j<span>++</span><span>)</span> <span>{</span>              
    <span>if</span> <span>(</span><span>(</span>j <span>&gt;=</span> emphasisPos<span>.</span>left <span>-</span> <span>1</span><span>)</span> <span>&amp;&amp;</span> <span>(</span>j <span>&lt;</span> emphasisPos<span>.</span>right <span>-</span><span>1</span><span>)</span><span>)</span> <span>{</span>
        
        att<span>[</span>j<span>]</span> <span>+=</span> <span>2.5</span><span>;</span> 
    <span>}</span>
<span>}</span></code></pre></div>
<p>Note that this is pre-softmax, so we still end up with a nice distribution in the attention tensor. I chose to add 2.5 by fiddling around a bit - any less and it was drowned out, any more and it broke the attention mechanism entirely. Let’s see an example! Here’s a generation with no emphasis:</p>
<p><span>
      <a href="https://www.seangoedecke.com/static/bd6c34a3b3ec10abaaea6753ce21a64c/387f2/no-emphasis.png" target="_blank" rel="noopener">
    <span></span>
  <img alt="no-emp" title="no-emp" src="https://www.seangoedecke.com/static/bd6c34a3b3ec10abaaea6753ce21a64c/fcda8/no-emphasis.png"></a>
    </span></p>
<p>Now let’s emphasize the word “crow”, in an attempt to get a more crow-focused story:</p>
<p><span>
      <a href="https://www.seangoedecke.com/static/5f434cb514700e985bcc128527238042/6bfbb/emphasis.png" target="_blank" rel="noopener">
    <span></span>
  <img alt="emp" title="emp" src="https://www.seangoedecke.com/static/5f434cb514700e985bcc128527238042/fcda8/emphasis.png"></a>
    </span></p>
<p>It worked! Now he’s flying, he’s in a more crow-appropriate scenario, and another crow shows up later on. Obviously this is an incredibly crude approach, but it’s interesting that even in this form it works more or less as you’d expect it to. I suspect there’s a lot of mileage to be had out of a more sophisticated attention-hacking mechanism in LLMs.</p>
<p>What haven’t I done yet? One thing I want to do that I haven’t done yet is to play around with parallelizing the model. Any performance lessons I learn aren’t likely to be relevant to actual scale LLMS, since my toy model will be dominated by bottlenecks that aren’t present in any productionized inference steps. Still, it’ll tell me what can be parallelized and what can’t. I know per-head attention can be trivially parallelized, and I want to try out the trick from this <a href="https://x.com/karpathy/status/1697318534555336961?s=46&amp;t=IkBfJ4QW2Yh1eU-QP0dAQQ">@karpathy tweet</a>:</p>
<ol>
<li>Take a slow model and a fast model</li>
<li>Generate a five-token sequence sequentially with the fast model</li>
<li>Take that five-token sequence and generate all those sub-sequences in parallel (e.g for abcd, generate a, b given a, c given ab, d given abc) with the slow model</li>
<li>If any of these slow model generations don’t pick the fast model’s token, throw it the fast model sequence away and keep generating slowly</li>
</ol>
<p>The idea here is that (a) parallel token processing is surprisingly performant for large models, since the big performance hit is packing the transformer weights into memory, and (b) most token generation is obvious and can be done by smaller models, such as completing words/adding spaces after full stops/etc.</p>
<p>In general I strongly recommend porting llama2.c yourself and fidding around. It’s been a huge learning experience for me.</p></section></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via www.seangoedecke.com https://ift.tt/XtiKa4W<br>
<br>
November 14, 2024 at 10:15AM
